pilot/cmd/pilot-agent/options/agent.go

```diff
@@ -71,6 +73,9 @@ func NewAgentOptions(proxy *model.Proxy, cfg *meshconfig.ProxyConfig) *istioagen
 		DualStack:                   features.EnableDualStack,
 		UseExternalWorkloadSDS:      useExternalWorkloadSDSEnv,
 		MetadataDiscovery:           enableWDSEnv,
+		SDSFactory: func(options *security.Options, workloadSecretCache security.SecretManager, pkpConf *meshconfig.PrivateKeyProvider) istioagent.SDSService {
+			return sds.NewServer(options, workloadSecretCache, pkpConf)
+		},
```

Comment:
I was wondering - can we refactor the code so that components like SDS are in their own package, and the agent uses conditional compilation to load them ? Envoy using SDS to get certs from Istiod is great, but the agent use of SDS has always been over-complex and not necessary.

If we move to a model where Sidecar is also sandwiched - ztunnel taking care of security - we can greatly simplify and reduce the code ( and risks ), while keeping most of the sidecar functionality and benefits.

Reply:
Go plugins are terrible, so dynamic module loading is unlikely to work. Wasm was suggested, but then these plugins are gRPC heavy which also don't work well in Wasm. My proposal was to just use files as SDS data, and then run the providers in a separate, optional process, which eventually would be replaced by https://github.com/kubernetes/enhancements/pull/4318